src/Servers/Kestrel/Core/src/KestrelServer.cs

```diff
@@ -211,73 +211,135 @@ public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            await _bindSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
 
-                for (int i = 0; i < connectionTransportCount; i++)
+            try
+            {
+                if (_stopping == 1)
                 {
-                    (IConnectionListener listener, Task acceptLoop) = _transports[i];
-                    tasks[i] = Task.WhenAll(listener.UnbindAsync(cancellationToken).AsTask(), acceptLoop);
+                    throw new InvalidOperationException("Kestrel has already been stopped.");
                 }
 
-                for (int i = connectionTransportCount; i < totalTransportCount; i++)
+                IChangeToken reloadToken = null;
+
+                if (Options.ConfigurationLoader?.ReloadOnChange == true && (!_serverAddresses.PreferHostingUrls || _serverAddresses.Addresses.Count == 0))
                 {
-                    (IMultiplexedConnectionListener listener, Task acceptLoop) = _multiplexedTransports[i - connectionTransportCount];
-                    tasks[i] = Task.WhenAll(listener.UnbindAsync(cancellationToken).AsTask(), acceptLoop);
+                    reloadToken = Options.ConfigurationLoader.Configuration.GetReloadToken();
                 }
 
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+                Options.ConfigurationLoader?.Load();
 
-                if (!await ConnectionManager.CloseAllConnectionsAsync(cancellationToken).ConfigureAwait(false))
-                {
-                    Trace.NotAllConnectionsClosedGracefully();
+                await AddressBinder.BindAsync(Options.ListenOptions, AddressBindContext).ConfigureAwait(false);
+                _configChangedRegistration = reloadToken?.RegisterChangeCallback(async state => await ((KestrelServer)state).RebindAsync(), this);
+            }
+            finally
+            {
+                _bindSemaphore.Release();
+            }
+        }
 
-                    if (!await ConnectionManager.AbortAllConnectionsAsync().ConfigureAwait(false))
-                    {
-                        Trace.NotAllConnectionsAborted();
-                    }
-                }
+        private async Task RebindAsync()
+        {
+            await _bindSemaphore.WaitAsync();
+
+            IChangeToken reloadToken = null;
 
-                for (int i = 0; i < connectionTransportCount; i++)
+            try
+            {
+                if (_stopping == 1)
                 {
-                    (IConnectionListener listener, Task acceptLoop) = _transports[i];
-                    tasks[i] = listener.DisposeAsync().AsTask();
+                    return;
                 }
 
-                for (int i = connectionTransportCount; i < totalTransportCount; i++)
+                reloadToken = Options.ConfigurationLoader.Configuration.GetReloadToken();
+                var (endpointsToStop, endpointsToStart) = Options.ConfigurationLoader.Reload();
```

Comment:
Should there be one more Debug log for the case where a config change was detected but no endpoints appeared to be different? That way you at least know the reload was attempted.

Reply:
When I was manually testing it with file-backed config, I was seeing the token fire twice per change, so it might be a little noisy.

Reply:
Yeah, file notifications do that... That's why we have diffing in the first place.